### ISSUE TITLE
limit(fibonacci(n + 1)/fibonacci(n), n, oo) does not give GoldenRatio

### ISSUE THREAD

**Issue by asmeurer** [sympy/sympy#10382]
```
In [84]: limit(fibonacci(n + 1)/fibonacci(n), n, oo)
Out[84]:
    fibonacci(n + 1)
lim ────────────────
n─→∞  fibonacci(n)
```

The answer should be `GoldenRatio`.

**Comment by asmeurer** [comment]
Calling rewrite(sqrt) on the expression first to get the closed-form expression for fibonacci that uses GoldenRatio results in `NotImplementedError: Result depends on the sign of sign(log(-1 + sqrt(5)) + I*pi)`.

At any rate, assumedly to make this work, one just needs to define the series expansion for fibonacci.

**Comment by meghana1995** [comment]
I would like to work on this. I have a few questions though.
The closed form of Binet's formula can be written as Fn = (a**n - b**n) / (a-b) where a = (sqrt(5) + 1)/2 , b = (1 - sqrt(5))/2.
So Limit(n->oo) Fn+1 / Fn = Limit(n->oo) (a**(n+1) - b**(n+1)) /  (a**(n) - b**(n))
Would rewriting (factoring) a**n - b**n as  (a-b)*(a**(n-1) + a**(n-2)*b + a**(b-3)*(b**2) .. ) aid in computing the limit ? If so, is there a method in sympy to factor this expression ?
Is there a particular series expansion you were referring to ?

**Comment by asmeurer** [comment]
@meghana1995 wrap your expressions with backticks (like \`code\`) so that GitHub doesn't try to render them as Markdown. See https://guides.github.com/features/mastering-markdown/

**Comment by asmeurer** [comment]
Anyway, the rewrite(sqrt) comment above refers to doing this:

```
In [7]: fibonacci(n).rewrite(sqrt)
Out[7]:
 -n    ⎛        n            n⎞
2  ⋅√5⋅⎝(1 + √5)  - (-√5 + 1) ⎠
───────────────────────────────
               5
```

I called this on the input expression and limit gave the NotImplementedError. I guess it doesn't like the exponent with the negative base. 

Apparently the series expansion has complex coefficients. I don't know if the gruntz algorithm can handle this. 

Side note: `fibonacci(n).rewrite(GoldenRatio)` ought to work. That is easy to fix.

**Comment by Nitin216** [comment]
@asmeurer Hey, is this issue done or not.

**Comment by asmeurer** [comment]
I don't think so. Feel free to work on it.

**Comment by leosartaj** [comment]
> Side note: fibonacci(n).rewrite(GoldenRatio) ought to work. That is easy to fix.

@asmeurer does ``rewrite`` accept a singleton?

**Comment by asmeurer** [comment]
No, it doesn't look like it works. rewrite expects a string or a class.  Perhaps for certain instances it should check the class of the instance. Maybe instances of NumberSymbol?

**Comment by asmeurer** [comment]
Or we could just do it for all instances. But rewrite(sin(x)) being the same as rewrite(sin) seems a bit odd, and it would be an unintended consequence.

**Comment by leosartaj** [comment]
> No, it doesn't look like it works. rewrite expects a string or a class. Perhaps for certain instances it should check the class of the instance. Maybe instances of NumberSymbol?

I think we can support instances of ``NumberSymbol``. This will be a special case, though. They are generally all singletons, right? By allowing this something like ``_eval_rewrite_as_GoldenRatio`` can be allowed. Presently, for solving the current issue. I see when can either define ``_eval_rewrite_as_tractable`` that returns in the form of ``GoldenRatio``. Or we could simply change ``_eval_rewrite_as_sqrt`` to give the desired result. ``_eval_rewrite_as_tractable`` will be simply ``_eval_rewrite_as_sqrt``in this case.

> Or we could just do it for all instances. But rewrite(sin(x)) being the same as rewrite(sin) seems a bit odd, and it would be an unintended consequence.

-1 for allowing all instances. We could allow singletons though.

**Comment by SagarB-97** [comment]
@asmeurer 
I would like to take this up and as a first step I would like to implement `rewrite(GoldenRatio)` for `fibonacci`.

But, `GoldenRatio` is a `Singleton` and the `rewrite()` function has not been implemented for `Singleton` objects.

So there are two ways it can be done.
1. I can add `try` and `except` blocks in `rewrite()` to handle `Singleton` objects.
2. `rewrite()` for `Singleton` s must be called as : 

```python
fibonacci(n).rewrite('GoldenRatio')
```

**Comment by SagarB-97** [comment]
The above PR implements it using method (1).

**Comment by SagarB-97** [comment]
Some consequences of changes : 

```python
>>> limit(fibonacci(n+1)/fibonacci(n),n,oo)
GoldenRatio
>>> limit(fibonacci(n)/fibonacci(n-1),n,oo)
GoldenRatio
>>> limit(fibonacci(n+2)/fibonacci(n),n,oo)
GoldenRatio**2
>>> limit(fibonacci(n+i)/fibonacci(n),n,oo)
GoldenRatio**i
```

**Comment by SagarB-97** [comment]
The above fix works for various definitions of `n` including 
```python
n = symbols('n')
```
and
```python
n = Symbol('n' , integer=True)
```

However, a `NotImplementedError()` is raised if `n` is defined as : 
```python
n = Symbol('n' , positive=True, integer=True)
```
or 
```python
n = Symbol('n' , positive = True)
```

I think this might be a bug in the `doit()` function. The expression `fibonacci(n+1)/fibonacci(n)` only gets converted to an expression containing `GoldenRatio` raised to various `n` powers.

**Comment by jksuom** [comment]
Closed by #12176.

**Comment by anutosh491** [comment]
@jksuom @asmeurer This should be reopened  !! 

Currently on master
```
>>> n = Symbol('n')
>>> limit(fibonacci(n + 1)/fibonacci(n), n, oo)
1
>>> gruntz(fibonacci(n + 1)/fibonacci(n), n, oo)
Traceback (most recent call last):
.....................
Asymptotic expansion of fibonacci around [oo] is not implemented.
```
This should be solved through a `_eval_rewrite_as_tractable` function where is it expressed in form of `GoldenRatio` and `sqrt`.
We can do that through
```
    def _eval_rewrite_as_tractable(self, n, **kwargs):
        from sympy.functions import sqrt, cos
        return (S.GoldenRatio**n - cos(S.Pi*n)/S.GoldenRatio**n)/sqrt(5)
```
After applying this the `gruntz` based limit would be fixed.
```
>>> gruntz(fibonacci(n + 1)/fibonacci(n), n, oo)
GoldenRatio
```
But the limit method still won't work. This is because
```
>>> fibonacci(1 + 1/n)/fibonacci(1/n)
fibonacci(1 + 1/n)/fibonacci(1/n)
>>> _.as_leading_term(n)
1
>>> fibonacci(1 + 1/n).as_leading_term(n)
fibonacci(1/n)
```
We get this from the last line here.
https://github.com/sympy/sympy/blob/d9aae45d682d97bf74ded0236768ec7adebbf9de/sympy/core/function.py#L834-L837

Here `args` is `[self.args[0].as_leading_term(n)]` which is `(1 + 1/n).as_leading_term(n) == [1/n]`. Then we return `self(*args)` which gives use `fibonacci(1/n)` . Which cancels out with the denominator term ! I am not sure what is the logic behind returning this and not `self` as self would give us `fibonacci(1 + 1/n)` which seems more appropriate (and returning `self` also passes all the tests just in case)
